[SPARK-54739][INFRA] Cancel workflow if a new commit is pushed

gaogaotiantian opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

We cancel the workflow if there is new commit pushed to the same branch.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

The existing workflow is super long (1~2hours). github limits the concurrent jobs for each repo, which means if we push a few commits in a short period of time, the latest commit will have to wait for the old ones to finish. However, in reality, we don't care about the result of the previous commits - they are irrelevant.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as new features, bug fixes, or other behavior changes. Documentation-only updates are not considered user-facing changes.

If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No, it's an infra change.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

It's not code change.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->

No.

gaogaotiantian committed:
Cancel workflow if a new commit is pushed

HyukjinKwon commented on .github/workflows/build_and_test.yml:
```diff
@@ -53,6 +53,9 @@ on:
         required: false
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
This actually was proposed before https://github.com/apache/spark/pull/40822. Do you remember why it was closed, @yaooqinn ?

gaogaotiantian replied:
It was correctly cancelled - https://github.com/gaogaotiantian/spark/runs/55141811395

gaogaotiantian committed:
Test CI to see if it works

gaogaotiantian committed:
Revert blank line change

gaogaotiantian committed:
Refactor concurrency group for build and test workflow

Updated concurrency group naming for workflows to handle master commits correctly.

gaogaotiantian committed:
test CI

gaogaotiantian committed:
Remove extra newline in build_and_test.yml

gaogaotiantian committed:
Update concurrency group for build and test workflow

gaogaotiantian committed:
Do not report test result if the workflow is cancelled

dongjoon-hyun reviewed (changes requested):
Sorry but I'm -1 especially during the Apache Spark 4.1.0 release prepartion. This will put me into very difficult situation because I need to verify every commit result, @gaogaotiantian , @HyukjinKwon , @yaooqinn .

Maybe, we can revisit this later.

gaogaotiantian commented on .github/workflows/build_and_test.yml:
```diff
@@ -53,6 +53,9 @@ on:
         required: false
+concurrency:
+  group: build-test-${{ github.workflow }}-${{ github.repository == 'apache/spark' && github.run_id || github.ref }}
+  cancel-in-progress: true
```
The current solution runs workflows for every commit in `apache/spark` in parallel - we put them into different groups. So if the commit is in `apache/spark`, they will be grouped by their `github.run_id`, which is unique for every run. Otherwise it will group everything and cancel the previous commit runs. @dongjoon-hyun do you need every commit result in forked repo?

HyukjinKwon replied:
ah ok makes sense

dongjoon-hyun commented:
Thank you so much for your understanding and paying attention to our INFRA, @gaogaotiantian . Ya, we can continue to discuss on this attempt.

dongjoon-hyun commented:
To prevent accidental merging, I marked this as `Draft` status and add `[WIP]` tag in PR title.

zhengruifeng commented on .github/workflows/build_and_test.yml:
```diff
@@ -53,6 +53,9 @@ on:
         required: false
+concurrency:
+  group: build-test-${{ github.workflow }}-${{ github.repository == 'apache/spark' && github.run_id || github.ref }}
+  cancel-in-progress: true
```
It seems we can follow the [example](https://docs.github.com/en/enterprise-cloud@latest/actions/how-tos/write-workflows/choose-when-workflows-run/control-workflow-concurrency#example-only-cancel-in-progress-jobs-on-specific-branches) with
```
cancel-in-progress: ${{ github.repository != 'apache/spark'}}
```

HyukjinKwon reviewed (approved):
I am fine with trying this out but probably let's just wait it out till 4.2 release. I will merge when that's done

gaogaotiantian commented:
JIRA is filed for this PR and I explained the background in that JIRA. This PR is ready for merge.

dongjoon-hyun reviewed (approved):
Thank you for waiting for Apache Spark 4.1.0. +1, LGTM.
